### DISCUSSION TITLE
Dreame L20 support?

### DISCUSSION THREAD

**Discussion by incline02** [merdok/homebridge-miot#549] [General]
Hi
tThanks for the incredible work already!
As I understand, Dreame L20 Ultra doesn't use the Mi app at all. Is there any chance your plugin may support it anyway?
thanks

**Comment by merdok** [comment]
Hi, I guess the plugin does support that the device. The only issue is on how to get the token from the device😅
When dreame switched to their own app, it is not possible anymore to retrieve the token from the xiaomi cloud. 
But as soon as you find a way to get the token and enter it in the config.json, it should work...

**Comment by incline02** [comment]
Thanks ! Have you had a look into it already?

Either way I’ll start looking.

**Reply by merdok** [reply]
I do not own a dreame device hence I did not look at it yet...

**Comment by incline02** [comment]
Ah crap, allow me one last question:

Can I help you have a look at it from the distance? Is it something feasible? I got MacOS, iOS, RaspOS with home bridge on it and the L20 Ultra of course.

**Reply by merdok** [reply]
Not really... I guess the only way would be to try to connect to the dreame cloud and get the token from there. But I will need to investigate it first...

**Comment by incline02** [comment]
Well, just as you answered, someone answered this that they’ll check out tomorrow if the token can be found via Home Assistant. https://www.reddit.com/r/Dreame_Tech/comments/16x7tlk/comment/kagsyz9/?context=3

To be continued… ! Thanks anyway for your quick answers

**Reply by merdok** [reply]
Sounds good, please let me know the result :)

**Reply by incline02** [reply]
Hey @merdok ! 
I didn't get no answers from the other Redditor but I've done my search, managed to open up one of my back ups and search the Dreame app repository and notably, the plist of the devices. I think I found what I'm looking for and I'm trying it as we speak. I only hid my account ID. 

anything interesting? 

<img width="1536" alt="image" src="https://github.com/merdok/homebridge-miot/assets/112760504/efab3f4e-34e1-4b7b-aeeb-0077f14c084b">

**Reply by merdok** [reply]
Does not seems like the token is included there. Just wondering if the dreame app goes over the cloud with commands or through local connection.

**Reply by incline02** [reply]
I've managed to have the switch in homekit but it fails. 

I used the highlighted string in the picture below. 
<img width="1425" alt="image" src="https://github.com/merdok/homebridge-miot/assets/112760504/2c1dd440-5d65-4f35-84b7-c9f926b76f7c">

and so the plugin returns in terminal's logs "device not supported, request support for the device" :D

**Comment by incline02** [comment]
Ive tested sending commands over 5G and it works so we got that covered.

In the meantime, that was one file out of a hundred. I’m going through them but evidently, they’ve changed their nomenclature and so when tutorials said to look for USERID or stuff in the MiCloud app. So basically, I’m testing all 32 digits with names closed to ID, Token or else.

I’ll report back but in the meantime, any suggestions?

**Reply by incline02** [reply]
Hey sorry for the spamming. Really. But I just found out: Dreame Home app has is also available on MacOS. Do you think I could find the token when I click on it and chose "Show Content Package"?

**Reply by merdok** [reply]
The thing is, when the app uses only the cloud to connect to the device then a token is not needed. You should try to control your device after disabling internet to check if that also still works.

**Comment by incline02** [comment]
Damn it. It seems to use only the cloud. Ive spent half the day trying to figure it out…….!

So there’s no chance your plugin can be updated to support L20?

**Comment by incline02** [comment]
Damn it. It seems to use only the cloud. Ive spent half the day trying to figure it out…….!

So there’s no chance your plugin can be updated to support L20?

**Reply by merdok** [reply]
The token is stored in the cloud, a connection to the cloud would be required to get the token from there, but currently the plugin does not support to connect to the dreame cloud.

**Reply by incline02** [reply]
and since you don't have a L20, you can't implement it right?

Can I help in any way? I'd be very happy to do it

**Reply by merdok** [reply]
No, i do not have a dreame device hence it is does not make it easier. The main problem here is that dreame switched from the xiaomi cloud to their own cloud...

**Comment by amatacchieri** [comment]
First of all, thank you @merdok, for your incredible work on this plugin.

I discovered this discussion after hours of trying to get the plugin working. I managed to obtain the token using the tcpdump/nc procedure (while connected to the adhoc wifi), but I can’t seem to get the switch to work (although it has been created). I’m not entirely sure I understand the cloud token issue and whether there’s any way I can contribute (I'm a developer and I own a L20 Ultra). Could you please help me clarify this or provide further details?
Many thanks!

**Reply by merdok** [reply]
So after you obtained the token, what is the issue which you are getting? Also are you trying to use a local connection or a micloud connection. I think your device cannot be added to the Xiaomi home app, right? In that case I do not think that the connection over micloud will work.

**Reply by amatacchieri** [reply]
hi @merdok,
so well I'm getting very different messages depending on what kind of configuration, most of the tries gave me:


```[18/08/2025, 11:24:50] [miot] [Roby] Error: Could not connect to device, handshake timeout
[18/08/2025, 11:24:50] [miot] [Roby] Error during device identify! Retrying in 60 seconds!
[18/08/2025, 11:25:50] [miot] [Roby] (Protocol) Start handshake 10.0.0.171```


The host is up, so the message above seems to be caused by an incorrect token. However, after several attempts, the plugin eventually managed to create a toggle in Home. I noticed that it only works when I set the Device ID not just under “Device ID” in the Global MiCloud Settings, but also under “Device Model” in the Advanced Settings. I used the model ```dreame.vacuum.r2228o
```

While trying to set this value also in "The device model", I get different logs and the switch created (though doesn't work):

```19/08/2025, 07:27:14] [miot] [Roby] Successfully created a RobotCleaner device! It is a Dreame L10s Ultra.
[19/08/2025, 07:27:14] [miot] [Roby] Initializing accessory!
[19/08/2025, 07:27:14] [miot] [Roby] Creating RobotCleaner accessory for device Roby!
[19/08/2025, 07:27:14] [miot] [Roby] Found accessory class at ../modules/robotcleaner/RobotCleanerAccessory.js!
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Creating Dock wrapper of type Property Monitor for property vacuum:status
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Initializing Dock wrapper of type Property Monitor for property vacuum:status
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Creating property monitor occupancy sensor!
[19/08/2025, 07:27:14] [miot] [Roby] Successfully created Dock wrapper of type Property Monitor!
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Creating Mode wrapper of type Property for property vacuum:mode
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Initializing Mode wrapper of type Property for property vacuum:mode
[19/08/2025, 07:27:14] [miot] [Roby] <-W-> Property type: Value List! Creating list item switches!
[19/08/2025, 07:27:14] [miot] [Roby] Successfully created Mode wrapper of type Property!
[19/08/2025, 07:27:14] [miot] [Roby] Accessory successfully initialized!
[19/08/2025, 07:27:14] [miot] [Roby] Final properties to monitor: [
  "vacuum:status",
  "vacuum:mode",
  "vacuum:fault",
  "battery:battery-level",
  "battery:charging-state",
  "brush-cleaner:brush-left-time",
  "brush-cleaner:brush-life-level",
  "filter:filter-life-level",
  "filter:filter-left-time",
  "vacuum-extend:cleaning-time",
  "vacuum-extend:cleaning-area",
  "clean-logs:total-clean-time",
  "clean-logs:total-clean-times",
  "clean-logs:total-clean-area"
]
[19/08/2025, 07:27:14] [miot] [Roby] Registering 1 accessories!
[19/08/2025, 07:27:14] [miot] [Roby] Everything looks good! Initiating property polling!
[19/08/2025, 07:27:14] [miot] [Roby] (Protocol) Start handshake 10.0.0.171
[19/08/2025, 07:27:14] [miot] [Roby] (Protocol) Server listening 0.0.0.0:38501
[19/08/2025, 07:28:19] [miot] [Roby] (Protocol) Start handshake 10.0.0.171
[19/08/2025, 07:28:24] [miot] [Roby] Error: Could not connect to device, handshake timeout
[19/08/2025, 07:28:24] [miot] [Roby] Error during connection! Retrying in 60 seconds!
```

so I'm a bit confused as it seems that the plugin is somehow able to login using the token, is my assumption right?

**Reply by merdok** [reply]
The token is only used for a local connection and in the case of a cloud connection to retrieve the unique device id.

Anyway, you should try to use the `miot test` cli command which this plugin also provides to verify that the token and ip are correct and that a device connection can be established.

As I already mentioned, if you are not able to add your device to the Xiaomi home app then a cloud connection will work and the only way would be a local connection which you can verify if it works with the cli command.

**Reply by amatacchieri** [reply]
Thanks a lot @merdok.

As suggested:

```
/var/lib/homebridge $ miot test -d --token 6VRC5f7aXPSyUmE3 10.0.0.171
INFO  Connecting to device at 10.0.0.171
Could not connect to device, handshake timeout
ERROR  Could not connect to the device at 10.0.0.171! Make sure the ip and token are correct!
 ```
 
 ```
nmap -sT 10.0.0.171

Starting Nmap 7.95 ( https://nmap.org ) at 2025-08-22 16:22 CEST
Nmap scan report for roby (10.0.0.171)
Host is up (0.0071s latency).
Not shown: 999 closed tcp ports (conn-refused)
PORT   STATE SERVICE
53/tcp open  domain

Nmap done: 1 IP address (1 host up) scanned in 3.97 seconds

 ```
 
 I suppose that the fact that the only port open is the 53 (DNS??) is not a good sign anyway, right?

**Reply by merdok** [reply]
Yeah the port is usually **54321** for a local connection. If the miot test command fails then the plugin will not work locally for sure...

**Reply by amatacchieri** [reply]
> Yeah the port is usually **54321** for a local connection. If the miot test command fails then the plugin will not work locally for sure...

I've corrected my command, those are the correct ports open on the device (I've missed to force a full tcp port):

```
Starting Nmap 7.95 ( https://nmap.org ) at 2025-08-30 17:15 CEST
Nmap scan report for dreame_vacuum_r2253w (10.1.0.136)
Host is up (0.012s latency).
Not shown: 65533 closed tcp ports (conn-refused)
PORT     STATE SERVICE
53/tcp   open  domain
8123/tcp open  polipo
```

**Reply by merdok** [reply]
These should not be used by the miot protocol

**Reply by amatacchieri** [reply]
> These should not be used by the miot protocol

thanks @merdok for you prompt reply. So I assume that the absence of the TCP port 54321 confirms we can't use it with homebridge (as the robot is communicating only via cloud), is my understanding correct?

**Reply by merdok** [reply]
It seems so, apparently the new Dreame models communicate differently.
